pages/DSNP/UserData.md

```diff
@@ -0,0 +1,84 @@
+  * Include the previous `etag` value for the chunk. If the chunk is new, `etag` should be set to `null`.
+  If any chunks are to be deleted, they should be included in the input identified with the existing `etag` and a `null` value for the data.
+
+If the Operation is invoked successfully it MUST (synchronously) return a new set of `etag` values for each data type replaced, corresponding to the updated state of the data (with new chunks added and deleted chunks removed).
+Applications should not interpret this response as an indication that the operation was completed and a state change record emitted, as this typically occurs asynchronously.
+However, this strategy allows applications to make the optimistic assumption that, in due course, the DSNP system will reflect the intended changes, without needing to wait for asynchronous confirmation.
+
+If, on the other hand, an invocation of Replace User Data is rejected due to entity tag discrepancies, this indicates that the relevant data has changed since the last preimage was acquired, and the application should fetch the most current version of the relevant data with the Get User Data Operation, reapply any intended changes, and retry the operation.
+
+The Replace User Data Operation MUST generate a [User Data Replaced Record](Records.md#user-data-replaced) containing the DSNP User Id and the set of updated user data types (but not the data itself).
+If the implementation detects that no change has occurred, it SHOULD omit the relevant unchanged data types from the state change record.
+
+# Get User Data Operation
+
+The Get User Data Operation takes the following parameters:
+* The DSNP User Id of the user who controls the data
+  * Note: While _writing_ user data is reserved for the user and any delegates, any authenticated principal can read any user's data (though it may be encrypted).
+* The user data types (by system name) that should be retrieved.
+
+The operation returns a mapping of user data type to data chunks and entity tags (the same general structure as the input data for Replace User Data), for each requested data type.
+If no chunks for a requested data type exist, an implementation MAY omit that data type from the response.
+
+To transform the public data from the output to Avro binary records, a consumer should apply the following algorithm to each data type included:
+* Determine the relevant encryption algorithm, compression codec, and object schema from the user data type and version noted.
+* For each chunk,
+  * If encryption is indicated, decrypt the chunk data using the user's secret key as in the [libsodium](https://doc.libsodium.org/public-key_cryptography/sealed_boxes) function `crypto_box_seal_open`.
+  * If compression is required, uncompress the chunk data using the specified codec.
+  * Append the chunk to the list of records for the data type.
```

Comment:
Can this happen before deserialization? Actually the deserialization is not specified here, so not sure which way you intended it to go.

Reply:
I will clarify ("Append the chunk" is not accurate anymore). The data after the uncompress step is a sequence of Avro binary objects, so can be deserialized into logical records after that.